File: Competitions/Begginer Contests/codPi/G.cpp
#include <bits/stdc++.h>
int DEBUG = 0, QQQ = 0;
using namespace std;

double getNorm(int p1[2], int p2[2])
{
  return(sqrt(pow(p2[1] - p1[1], 2) + pow(p2[0] - p1[0], 2)));
}

double getAngle(int p1[2], int p2[2])
{
  if (atan2(abs(p2[1] - p1[1]), abs(p2[0] - p1[0])) == 0) return(round(1000000 * 2 * atan(1)) / 1000000);
  return(round(1000000 * abs(atan2(abs(p2[1] - p1[1]), abs(p2[0] - p1[0])))) / 1000000);
}

int main()
{
  int tests; scanf("%d", &tests);
  while (tests > 0)
  {
    int size; scanf("%d", &size);
    int coordinates[size][2];
    for (int i = 0; i < size; i ++)
      scanf("%d %d", &coordinates[i][0], &coordinates[i][1]);

    map<double, vector<double> > mostAppearing;
    double theAngle = (size - 2) * (180/size), theNorm = 0;
    theAngle = round(1000000 * (theAngle * M_PI / 180.0)) / 1000000;
    if (DEBUG) printf("The angle %lf\n", theAngle);
    int isColinear = 0, isNotSameSize = 0;
    for (int i = 0; i < size; i ++)
    {
      map<double, int> colinear;
      for (int j = 0; j < size; j ++)
      {
        if (i != j)
        {
          if (mostAppearing.count(getAngle(coordinates[i], coordinates[j])) == 0)
            mostAppearing.insert({getAngle(coordinates[i], coordinates[j]), {0, getAngle(coordinates[i], coordinates[j]), getNorm(coordinates[i], coordinates[j])}});
          //if (theAngle == getAngle(coordinates[i], coordinates[j])) printf("%lf %lf\n", mostAppearing[getAngle(coordinates[i], coordinates[j])][2], getNorm(coordinates[i], coordinates[j]));
          if (theAngle == getAngle(coordinates[i], coordinates[j]) && mostAppearing[getAngle(coordinates[i], coordinates[j])][2] != getNorm(coordinates[i], coordinates[j]))
          {
            //printf("HOLYFOCK %lf %lf\n", mostAppearing[getAngle(coordinates[i], coordinates[j])][2], getNorm(coordinates[i], coordinates[j]));
            isNotSameSize = 1;
            break;
          }
          mostAppearing[getAngle(coordinates[i], coordinates[j])][0] ++;
          if (colinear.count(getAngle(coordinates[i], coordinates[j])) == 0)
            colinear.insert({getAngle(coordinates[i], coordinates[j]), 0});
          else
          {
            //isColinear = 1;
            //break;
          }
        }
      }
    }
    if (DEBUG)
    {
      for (auto i : mostAppearing)
        printf("%lf %lf %lf\n", i.second[0], i.second[1], i.second[2]);
    }
    if (isColinear)
    {
      if (DEBUG) printf("Colinear\n");
      printf("NO\n");
    }
    else
    {
      if (isNotSameSize || (mostAppearing.count(theAngle) == 0)) // || (mostAppearing.count(theAngle) > 0 and mostAppearing[theAngle][0] != size))
      {
        if (DEBUG) printf("Sides diff\n");
        printf("NO\n");
      }
      else
        printf("YES\n");
    }
    tests --;
  }
  return(0);
}

File: Competitions/Begginer Contests/codPi/G.py
from math import atan2, atan, degrees, sqrt

def norm(p1, p2):
    return(sqrt((p2[1] - p1[1])**2 + (p2[0] - p1[0])**2))
def ang(p1, p2):
    return(abs(atan2(p2[1] - p1[1], p2[0] - p1[0])))

tests = int(input())
while (tests > 0):
    #print()
    size = int(input())
    coordinates = []
    for i in range(size):
        coordinates += [list(map(int, input().split()))]
    #print(coordinates)
    mostAppearing, isColinear = {}, 0
    for i in range(size):
        colinear = {}
        for j in range(size):
            if (i != j):
                theAngle = ang(coordinates[i], coordinates[j])
                if (theAngle not in mostAppearing):
                    mostAppearing[theAngle] = [0, theAngle, norm(coordinates[0], coordinates[1])]
                mostAppearing[theAngle][0] += 1
                if (theAngle not in colinear):
                    colinear[theAngle] = 0
                else:
                    isColinear = 1
                    break
    if (isColinear == 1):
        #print("Colinear")
        print("NO")
    else:
        done = [0] * size
        appeared, bestAngle, bestDistance = 0, 180, 0
        for i in mostAppearing:
            if mostAppearing[i][0] > appeared:
                bestAngle, bestDistance = mostAppearing[i][1:]
                appeared = mostAppearing[i][0]
        regular = 0
        deg = degrees(bestAngle)
        #print("selected", degrees(bestAngle), bestDistance)
        if (deg != 180):
            for p in range(size):
                regular = 0
                if (done[p] < 4):
                    for i in range(size):
                        if (done[i] < 4):
                            if (bestAngle == ang(coordinates[p], coordinates[i]) and bestDistance == norm(coordinates[p], coordinates[i])):
                                #print("Wee", p, i)
                                done[i] += 1
                                done[p] += 1
                                regular = 1
                                break
                if (regular == 0):
                    break
        #print(colinear)
        if (regular == 1):
            print("YES")
        else:
            print("NO")
    tests -= 1